### DISCUSSION TITLE
Generate JSON graph

### DISCUSSION THREAD

**Discussion by stereobooster** [terrastruct/d2#2224] [Ideas]
Idea: option to generate graph in JSON format.

1. Than this data can be embedded in HTML `data-graph=...`
2. Than it can be read from JS and fed to graph lib, for example https://github.com/dagrejs/graphlib
3. And we can add interactivity with JS

Examples: https://beoe.stereobooster.com/examples/vizdom/.
Implementation: https://github.com/stereobooster/beoe/blob/main/packages/demo/src/components/vizdom.ts

This can be alternative approach to animation. Related:

- https://github.com/terrastruct/d2/issues/1461
- https://github.com/terrastruct/d2/issues/1485

**Comment by alixander** [comment]
This is a good idea for a function in the Javascript version

**Comment by stereobooster** [comment]
I implemented prototype https://github.com/stereobooster/beoe/pull/9 ([preview](https://deploy-preview-9--beoe.netlify.app/diagrams/d2/))

<img width="701" alt="Screenshot 2025-01-18 at 00 22 36" src="https://github.com/user-attachments/assets/e998751d-b6c4-4883-a504-5974dcd43e74" />

(cursor is over `c`, but not over text)

There are several problems though:
- edges have strange ids, like `(b -> c)[0]`. I need to use `CSS.escape` in order to query DOM by id
- it would be easier to work with SVG if all elements (edges, nodes) would have classes (in the same element where id)
  ```svg
  <g id="e"><g class="shape">...</g>
  ```
- ~~structure for edges is inconsistent~~ never mind this is due to SVGO
   ```svg
   <g id="(a -> b)[0]">
     <marker ...></marker>
     <path stroke-width="2" d="m0 0 10 6-10 6z" class="connection" ...></path>
   </g>
   <path id="(c -> d)[0]" fill="none" stroke="#0D32B2" marker-end="url(#mk-3488378134)" d="M361 33h94" class="connection stroke-B1" mask="url(#d2-424925946)" style="stroke-width:2"></path>
   ```

**UPD**

here is better link for example https://beoe.stereobooster.com/start-here/interactivity/#d2-example

**Reply by alixander** [reply]
related: https://github.com/terrastruct/d2/issues/2273

I'll prioritize that
